pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	utilerrors "k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	sentinelFileName = "..sentinel"
+	tsFormat         = "1981_02_01_15_04_05"
+)
+
+func ts() string {
+	return time.Now().Format(tsFormat)
+}
+
+// TODO: inject coordinates for logging traceability to pod/volume
+
+// AtomicWriter handles writing a payload to a target directory atomically.
+type AtomicWriter struct {
+	TargetDir       string
+	IdentifierToken string
+	LogContext      string
+}
+
+// currentStateDirName returns the name of the directory within the writer's
+// target directory that should hold the current projected state.
+func (w *AtomicWriter) currentStateDirName() string {
+	return fmt.Sprintf("..%v", w.IdentifierToken)
```

Comment:
%v is usually my go-to for not having to think, but i will switch to %s where necessary